types/underscore/index.d.ts

```diff
@@ -3710,9 +3711,13 @@ declare module _ {
         * @keys The key/value pairs to keep on `object`.
         * @return Copy of `object` with only the `keys` properties.
         **/
+        pick<T, K extends keyof T>(obj: T, ...keys: K[]): Pick<T, K>;
```

Comment:
I'm missing two signatures from [the documentation](https://underscorejs.org/#pick):

```ts
pick<T, K extends keyof T>(obj: T, keys: K[]): Pick<T, K>;
pick<T, K extends keyof T, V extends typeof T[K]>(obj: T, predicate: ObjectIterator<V, boolean>): Pick<T, K>;
```

In the second line, I'm not 100% sure about the notation of the third type parameter (`V extends typeof T[K]`), but I'm sure such a signature needs to be there because `pick` accepts a predicate.

Reply:
Hi,

The reason I didn't add the first one is that when I tested it, TypeScript failed to infer K automatically so I decided to drop it - but now I realize I was wrong... 
Same goes for the 2nd signature - TypeScript can't infer K by itslef but you are right we should at least enforce the predicate and return types.

So, I added the 2 missing signatures. but regarding the predicate signature - I changed it a bit and created a new `ObjectIterator` called `FullObjectIterator` as the object values doesn't have to be the same for all keys (like in the tests). I'm not sure about the name I gave it so please let me know your feedback on the matter.

Reply:
In my opinion, this is an improvement, but I think `T[K]` and the `ObjectIterator` interface work a bit differently than you seem to assume. If I'm right, I want to request a few more changes.

Let's say we have the following `T`.

```ts
interface T {
    apple: string;
    banana: number;
    cherry: boolean;
}
```

Then I think we agree that `keyof T` is `'apple' | 'banana' | 'cherry'`. I think we also agree that `T['apple']` is `string`. However, you seem to have different expectations about `T[keyof T]`. It expands to `T['apple' | 'banana' | 'cherry']`, so I think it will be `string | number | boolean`. The typescript handbook is not 100% clear on this situation, however.

Furthermore, I think `T` is a subset of `Dictionary<string | number | boolean>`. Note that the element values don't need to be the same for all keys in such a `Dictionary`, either. This would mean that any `ObjectIterator` that accepts `string | number | boolean` as its first argument will be able to iterate over `T`. In this case, you don't need `FullObjectIterator`, although `FullObjectIterator` is arguably slightly more correct than `ObjectIterator`.

Replacing `ObjectIterator`, or giving it a slightly different sibling, doesn't seem like a good idea to me. So I'd prefer that you stick closer to the predicate signature I suggested and just use `ObjectIterator`.

In `FullObjectIterator`, you wrote `K extends keyof T = keyof T` to provide a default when TS can't infer `K`. You can apply this trick to the `pick` function signatures too, if necessary.

Reply:
Thanks for the detailed response!
I agree with you completely! I removed my new predicate interface and reverted to use the already existing one.
As to your last suggestion - I don't it applies to the pick method as it doesn't make sense to perform a pick on all the keys of an object as it will act like a `copy` method, so I think it is better to let the user choose the subset of keys to pick.